### ISSUE TITLE
Figure out: relative URLs, page files, URL Rewriting etc.

### ISSUE THREAD

**Issue by fvsch** [fvsch/kirby-staticbuilder#14]
[Epic]
## Challenges
- When a page links to another page, will that link work with what we output?
- When a page references an asset, does that reference work with what we output?
- When a page references a page file (e.g. from Kirbytext), does that reference work with what we output?
- Can a static build be deployed at the root of any domain or subdomain?
- Can a static build be deployed as a subfolder (at any level)?
- Can a static build be distributed without a HTTP server (e.g. documentation pages, open in browser with `file://` protocol)?
## Current situation
1. Domain name in URLs depends on Kirby config. By default it will output the domain name of the request. We probably want to force `$kirby->urls()->index = '';` (limited but decent default), and optionally let users set the root URL to whatever they want.
2. Right now we're building pages with a suffix added to their URI, e.g. `my/page/index.html` or `my/page.html`. Depending on server config this might break URLs between pages. We should pick a default that works in most cases. I originally picked a `/index.html` suffix because my personnal site and main test project uses hardcoded relative references to files, e.g. `<img src="illustration.jpg">`. This might be an outlier case, and maybe `.html` is a better default suffix.
3. We copy over the `assets` and `thumbs` folder. This should be alright (and can be overridden if needed).
4. We copy over the `content` folder, so that file references from kirbytags or using `$page->image('some-image.jpg')->url()` keep working. That means we copy folders and source content for all pages, including some that might be marked as private or that was filtered by default or user-configured filter. Not awful but not great.
## Possible improvements

For the content folder: perhaps parsing the generated HTML for every page, identifying references to files in `content/…`, replacing with a different URL, and adding old and new URL to a map of assets to copy. Or is that overkill? Note that if we’re already copying over the page’s files, so we could rewrite only those known URLs.

If we want to go with some rewriting of the HTML output, we could forcefully set something like `$kirby->urls()->index = 'STATIC_URLS_INDEX'` then rewrite all the `STATIC_URLS_INDEX/…` as needed. This would allow us to rewrite URLs generated by Kirby (for known pages, assets, thumbs etc.) and not other paths or path-like strings in the output, avoiding a lot of opportunities for false positives.

We could _also_ add a generic rewriting config. If we do that, I favor using a callback: get a string (rendered page), do whatever you need, return a string. BUT the important part is that the default config and features of this plugin should make writing such a callback unnecessary for most users.
## Fully relative paths

Last possible improvement answers the "distribute without a HTTP server" case. This means making all file references relative:

``` html
<!-- Example: current page is static/category2/sub-section/page-slug.html -->

<link rel="stylesheet" href="../../assets/css/styles.css">

<!-- Nav -->
<a href="../../index.html"> Home </a>
<a href="../../category1.html"> Category 1 </a>
<a href="../../category2.html"> Category 2 </a>

<!-- Breadcrumbs -->
<a href="../../index.html"> Home </a>
<a href="../../category2.html"> Category 2 </a
<a href="../sub-section.html"> Sub-section </a>

<img src="page-slug/cover.jpg" alt="Image for this page">
```

I think it’s doable but this one clearly requires rewriting most URLs and file references.

This probably should be an option. Maybe using a single config key?
- `c::set('plugin.staticbuilder.urlbase', '/foo');` to set URLs starting with `/foo/` (e.g. `/foo/index.html` for home, `/foo/assets`, etc.)
- `c::set('plugin.staticbuilder.urlbase', 'https://somedomain.tld');` to set URLs starting with `https://somedomain.tld/`
- `c::set('plugin.staticbuilder.urlbase', false);` to use relative URLs instead.

**Comment by mogelbrod** [comment]
Just implemented URL rewriting with support for both absolute and relative URLs in mogelbrod/kirby-staticbuilder@14fd8092c676341fe425dc0a34e84fcefd9aaf7a .

Possible values for the `plugin.staticbuilder.urlbase` option:
- `false`: Relative URLs without /index.htm, meant to be served via server.
- `'file://'`: Relative URLs with /index.htm suffix appended to all URLs lacking an extension, can be served via the file system.
- `''`: Absolute URLs without domain.
- `'https://domain.tld'`: Fully qualified URLs.

All URLs generated by Kirbys helpers (`url()` for example) should be automatically rewritten. Links generated using standard markdown syntax are currently not rewritten, but kirbytags should work.

**Comment by fvsch** [comment]
That looks great.

Maybe I’m suffering from acute Not Invented Here syndrome, but since I had already started working on this I’d like to finish my implementation separately, perhaps stealing entire pieces from yours. Would that be okay?

**Comment by fvsch** [comment]
So I did something very similar in https://github.com/fvsch/kirby-staticbuilder/commit/9a9604978a0648b443f20ee092df00d19e34b035
I even wrote my own function for resolving the relative URLs, but yours was more efficient so I poached it. ^^

Some differences:
- option for base URL has just one special value: `'.'` for relative URLs.
- The option to add the extension/suffix for HTML page URLs is a separate option. It’s also implemented with an array of known URLs, rather than with a regexp. I wanted to limit regexp use as much as possible.
- When I do use a regexp, I did it a bit differently.
- I don’t restore the normal values for $kirby->urls etc. after I’m done. Not sure there’s a point.

Other than that it’s rather similar.

**Comment by mogelbrod** [comment]
Feel free to grab any code I've committed :)

I thought about only replacing known URLs first but ended up implementing a more liberal rewriter to ensure that dynamic routes (containing url params) and other edge cases are also replaced. Regex replacement is obviously not ideal, but since it requires the URL placeholder (which is pretty much guaranteed to only exist within a URL that should be replaced) I think my version is robust enough and avoids the added complexity of the url map.

I would suggest refactoring the `renderPage($page)` method into `rewriteUrls($text, $urlBase)` to enable rewriting any content (routes for example).

I don't think there's any point in restoring the url configuration afterwards, that part of the code was added just in case :)

**Comment by fvsch** [comment]
> I would suggest refactoring the renderPage($page) method into rewriteUrls($text, $urlBase) to enable rewriting any content (routes for example).

Makes sense. I had missed the route use case. ^^
Done in https://github.com/fvsch/kirby-staticbuilder/commit/97fc8bb998f6ba59b33bd870fd93b40c1c56adfe

**Comment by fvsch** [comment]
> I thought about only replacing known URLs first but ended up implementing a more liberal rewriter to ensure that dynamic routes (containing url params) and other edge cases are also replaced.

Good point. I moved to regex-only for ugly URLs (aka explicit URLs) in https://github.com/fvsch/kirby-staticbuilder/commit/33c20c31e1338164ae808583696521cc421d520e. Could you perhaps take a look at the `rewriteUrls` method and tell me if you see any obvious issues?

I’m using a different regexp because I wanted to be able to catch `"mycool-url'with'single-quotes"` (and the other way round, and be more restrictive for unquoted URLs. Maybe that’s totally unnecessary, but I find it easier to debug two separate patterns instead of having both patterns expressed with conditions/lookaheads/etc.

**Comment by fvsch** [comment]
Barring any bugs, to do for the 1.0 release: document it.

**Comment by mogelbrod** [comment]
I totally forgot about the fact that single quotes are valid in URLs. Your regex doesn't replace URLs present outside attributes ("free" URLs) however, so I suggest a few tweaks to fix that:

```
From: (["'])\s*(URLPREFIX[^<>]*?)\s*\1|(=)(URLPREFIX[^\s<>'"]*)
To:   (['"])URLPREFIX\/?([^">{}\s]*)\1|URLPREFIX\/?([^'">{}\s]*)
```

The tweaked regex allows single quotes if the URL is wrapped with double quotes, but not otherwise. Try it out at [regex101](https://regex101.com/r/eC7pB1/2).

Other than that my only comments on `rewriteUrls` are:
- preg_quote `static::URLPREFIX` before using it inside a regex, just in case it would be changed to include any magic characters.
- Replace `array_key_exists('extension', pathinfo($url))` with `pathinfo($url, PATHINFO_EXTENSION)`
- `$this` if AFAIK available inside closures, so if you use `$this->uglyurls` instead of `$uglyurls` you won't have to import it via `use()`.

**Comment by fvsch** [comment]
I implemented your three comments, but for the regex itself for now I’m keeping it more focused on the “HTML attribute or quoted URL” use case, NOT matching “free” URLs.

Main reason is that I want to be able to match valid and common HTML code like this:

``` html
<img src="
  URLPREFIX/my cool image.jpg
     ">
```

But maybe that’s the wrong approach. I’m thinking it would yield less false positives but maybe that’s naive.

**Comment by mogelbrod** [comment]
Nice!

URLs [can't contain unencoded spaces](http://stackoverflow.com/a/5442823).

I can't think of any possible false positives in the more liberal regex variant. Shouldn't the rewriter take care of all URLs outputted using kirbys URL helper? Attempting to limit replacements to only those present inside HTML-like attributes will prevent the following:
- Cases where a developer or user have triggered a call to `url($to)` in plain text, possibly when outputting the URL of the page in a printable version of it.
- Pages using templates whose output is a format different from HTML, for example manually rendering YAML:

``` php
<?php header::contentType('application/x-yaml') ?>
page:
  url: <?php echo $page->url() ?>
```

**Comment by fvsch** [comment]
Yeah I was looking at this again and we’re coming to similar conclusions.

The URLs we’re concerned with are output by Kirby’s helpers and methods, and they do encode spaces (disallowed in URLs but browsers still use them, so pages can work with illegal URLs) and single quotes (allowed in URLs). So we can break on whitespace, and I don’t think we need to test for quotes before and after the URL. I’m going to simplify the regex it a lot.

Why do you break on `{}` characters by the way?
(Thanks for your feedback!)

**Comment by mogelbrod** [comment]
Oh, kirbys helpers actually encode single quotes? That sure is useful!
I included the curly brackets since they're pretty much the only common delimeters [not allowed in a URL](http://stackoverflow.com/a/1547940/1527562). There will probably not be many cases where they're useful, but since they don't really add any complexity I left them in. You should probably also add `<`, `|` and `^` to the character classes that break URLs.

**Comment by fvsch** [comment]
Okay then, I’m going for `URLPREFIX\/?[^<>{}"'\s]*` and that’s about it.

**Comment by fvsch** [comment]
There are cases where the `relativeUrl` method gives the wrong results. It tends to go back one step too much, and when you want to do this:

```
from:  some/section/subsection
to:    some/section/subsection/page
```

it doesn't work because we need to output `subsection/page` (or `./subsection/page`) but we didn't save the `subsection` part.

**Comment by mogelbrod** [comment]
It shouldn't be a problem if using `/index.htm` as the suffix since the browser will end up at `some/section/subsection/` (trailing slash) when requesting `some/section/subsection` - this is only an issue if the suffix is changed to something like `file.htm`, right? The only other case where the result of `relativeUrl` has to be modified is when `$from = '/'` afaik ([part of my implementation which handles it](https://github.com/mogelbrod/kirby-staticbuilder/blob/master/core/builder.php#L454)).

**Comment by fvsch** [comment]
I ran a lot of tests and had a bunch of issues. Maybe I’m not abstracting it enough, but after testing a lot of things I ended up with a solution that is not _too_ awful and verbose. ^^

I think it should be fixed in https://github.com/fvsch/kirby-staticbuilder/commit/7d2696b8be51c7a840023d3f46dd3d3c92c359ea.
I didn’t reuse your approach for handling special cases because I wasn’t exactly sure how it worked :D which is probably a testament to my limits as a PHP dev. ^^

Closing this issue because everything discussed was implemented and it seems to finally work alright. If I find more bugs I will open more specific issues for them.
